File: src/hilbert/nodes/writer.py
"""Writer node for Hilbert — inline citations + enriched provenance."""

import json
import uuid
from datetime import datetime
from typing import Dict, List, Optional

from hilbert.config.settings import get_settings
from hilbert.llm import get_client, get_writer_prompt
from hilbert.models import Paper, Report
from hilbert.state.research import ResearchState


def _build_source_index(papers: List[Paper]) -> tuple[List[dict], Dict[str, str]]:
    """Build a 1-based numbered source list and a paper_id → citation number map."""
    source_data: List[dict] = []
    source_index: Dict[str, str] = {}  # paper_id → "1", "2", ...

    for i, paper in enumerate(papers[:20], start=1):
        num = str(i)
        source_data.append({
            "citation_number": num,
            "paper_id": paper.paper_id,
            "title": paper.title,
            "authors": [a.name for a in paper.authors],
            "published_date": str(paper.published_date) if paper.published_date else "n.d.",
            "url": str(paper.url) if paper.url else "",
            "doi": paper.doi or "",
            "venue": paper.venue or "",
        })
        source_index[paper.paper_id] = num

    return source_data, source_index


async def writer_node(state: ResearchState) -> dict:
    """Generate final research report with inline [N] citations."""
    settings = get_settings()

    query = state["query"]
    findings = state.get("findings", [])
    papers = state.get("papers", [])
    started_at: Optional[datetime] = state.get("started_at")
    gaps = state.get("gaps", [])
    contradictions = state.get("contradictions", [])

    callback = state.get("progress_callback")
    if callback:
        callback("writer", {"status": "writing"})

    source_data, source_index = _build_source_index(papers)

    findings_data = []
    for f in findings:
        findings_data.append({
            "finding_id": f.finding_id,
            "claim": f.claim,
            "confidence": f.confidence,
            "confidence_label": f.confidence_label(),
            "is_verified": f.is_verified,
            "source_paper_id": getattr(f, "source_paper_id", None),
            # Provide citation number directly so LLM can embed [N]
            "citation_number": source_index.get(getattr(f, "source_paper_id", ""), "?"),
        })

    system_prompt, user_prompt = get_writer_prompt(
        query=query,
        findings=findings_data,
        sources=source_data,
    )

    try:
        client = get_client()
        content = await client.complete_text(
            prompt=user_prompt,
            system_prompt=system_prompt,
            max_tokens=4000,
        )
    except Exception:
        content = generate_fallback_report(query, findings, papers)

    bibtex = generate_bibtex(papers)

    # Build contradiction section if any were confirmed
    contradictions_md = ""
    if contradictions:
        lines = ["## Contradictions Detected\n",
                 "> The following claim pairs were identified as potentially contradicting each other.\n"]
        for c in contradictions:
            lines.append(
                f"- **{c.severity_label()}** — *\"{c.claim_a}\"* vs *\"{c.claim_b}\"*"
                + (f"\n  {c.description}" if c.description else "")
            )
        contradictions_md = "\n".join(lines)

    sections = {"summary": content}
    if contradictions_md:
        sections["contradictions"] = contradictions_md

    report = Report(
        report_id=f"report-{uuid.uuid4().hex[:8]}",
        title=query,
        query=query,
        executive_summary="",
        sections=sections,
        sources=source_data,
        source_index=source_index,
        findings_summary=findings_data,
        bibliography=bibtex,
        created_at=datetime.now(),
        status="final",
    )

    output_dir = settings.output_dir
    output_dir.mkdir(parents=True, exist_ok=True)

    (output_dir / "report.md").write_text(report.to_markdown())
    (output_dir / "report.json").write_text(json.dumps(report.to_json(), indent=2))
    (output_dir / "report.bib").write_text(bibtex)

    provenance = generate_provenance(state, source_data, findings_data, gaps, started_at, contradictions)
    (output_dir / "report.provenance.md").write_text(provenance)

    return {
        "report": report,
        "status": "done",
    }


def generate_provenance(
    state: ResearchState,
    source_data: List[dict],
    findings_data: List[dict],
    gaps: List,
    started_at: Optional[datetime],
    contradictions: List = None,
) -> str:
    """Generate enriched provenance sidecar."""
    from hilbert.ui.mermaid import generate_research_mermaid
    from hilbert.config.settings import get_settings

    settings = get_settings()
    now = datetime.now()
    duration_s = int((now - started_at).total_seconds()) if started_at else 0

    verified_count = sum(1 for f in findings_data if f.get("is_verified"))
    fatal_gaps = [g for g in gaps if getattr(g, "severity", "") == "FATAL"]
    major_gaps = [g for g in gaps if getattr(g, "severity", "") == "MAJOR"]
    confirmed_contradictions = list(contradictions or [])

    dimensions = state.get("research_dimensions", [])
    dim_labels = [d.get("label", d.get("strategy", "?")) for d in dimensions]

    lines = [
        f"# Provenance: {state['query']}",
        "",
        "## Run metadata",
        f"- **Date:** {now.strftime('%Y-%m-%d %H:%M:%S')}",
        f"- **Duration:** {duration_s}s",
        f"- **Model:** {settings.model}",
        f"- **Rounds completed:** {state.get('round', 1)} / {state['max_rounds']}"
        + (" *(converged early)*" if state.get("converged") else ""),
        f"- **Research dimensions:** {', '.join(dim_labels) or 'default'}",
        "",
        "## Source pipeline",
        f"- **Sources consulted (pre-dedup):** {state.get('papers_consulted', len(source_data))}",
        f"- **Sources accepted (post-filter):** {len(source_data)}",
        "",
        "## Findings",
        f"- **Total findings:** {len(findings_data)}",
        f"- **Verified (confidence ≥ {settings.confidence_threshold}):** {verified_count}",
        f"- **Unverified:** {len(findings_data) - verified_count}",
        "",
        "## Reviewer integrity check",
        f"- **FATAL gaps:** {len(fatal_gaps)}",
        f"- **MAJOR gaps:** {len(major_gaps)}",
        f"- **Confirmed contradictions:** {len(confirmed_contradictions)}",
    ]

    if confirmed_contradictions:
        lines.append("")
        lines.append("### Contradictions")
        for c in confirmed_contradictions:
            lines.append(f"- [{c.severity_label()}] {c.description or 'No description'}")

    if fatal_gaps or major_gaps:
        lines.append("")
        lines.append("### Gap details")
        for g in fatal_gaps + major_gaps:
            lines.append(f"- [{g.severity}] {g.description}")

    lines.extend([
        "",
        "## Sub-questions",
    ])
    for sq in state.get("sub_questions", []):
        lines.append(f"- {sq}")

    lines.extend([
        "",
        "## Workflow",
        generate_research_mermaid(),
    ])

    return "\n".join(lines)


def generate_fallback_report(query: str, findings: List, papers: List) -> str:
    """Generate fallback report if LLM fails."""
    lines = [
        f"# Research Report: {query}",
        "",
        "## Summary",
        f"Found {len(findings)} findings from {len(papers)} papers.",
    ]

    for f in findings:
        lines.append(f"- {f.claim}")

    return "\n".join(lines)


def generate_bibtex(papers: List[Paper]) -> str:
    """Generate BibTeX bibliography."""
    lines = []
    for paper in papers:
        bib = paper.to_bibtex()
        if bib:
            lines.append(bib)
            lines.append("")
    return "\n".join(lines)


def create_writer_node():
    """Create writer node function."""
    return writer_node

File: cli/src/core/nodes/writer.ts
import * as fs from "fs";
import * as path from "path";
import { getSettings } from "../config";
import { getClient } from "../llm/client";
import { getWriterPrompt } from "../llm/prompts";
import { toMarkdown, toJson, toLatex } from "../models/report";
import { citationKey, toBibtex } from "../models/paper";
import { confidenceLabel } from "../models/finding";
import type { ResearchState } from "../state";
import type { Paper } from "../models/paper";
import type { Gap } from "../models/finding";
import type { Contradiction } from "../models/finding";

function buildSourceIndex(papers: Paper[]): [{ paper_id: string; title: string; authors: string[]; published_date: string; url: string; doi: string; venue: string; citation_number: string }[], Record<string, string>] {
  const sourceData: { paper_id: string; title: string; authors: string[]; published_date: string; url: string; doi: string; venue: string; citation_number: string }[] = [];
  const sourceIndex: Record<string, string> = {};

  for (let i = 0; i < Math.min(papers.length, 20); i++) {
    const paper = papers[i];
    const num = String(i + 1);
    sourceData.push({
      citation_number: num,
      paper_id: paper.paper_id,
      title: paper.title,
      authors: paper.authors.map((a) => a.name),
      published_date: paper.published_date || "n.d.",
      url: paper.url,
      doi: paper.doi || "",
      venue: paper.venue || "",
    });
    sourceIndex[paper.paper_id] = num;
  }

  return [sourceData, sourceIndex];
}

function generateFallbackReport(query: string, findings: { claim: string }[], papers: Paper[]): string {
  const lines = [
    `# Research Report: ${query}`,
    "",
    `## Summary`,
    `Found ${findings.length} findings from ${papers.length} papers.`,
    "",
    "## Findings",
  ];

  for (const f of findings) {
    lines.push(`- ${f.claim}`);
  }

  return lines.join("\n");
}

function generateBibtex(papers: Paper[]): string {
  return papers.map((p) => toBibtex(p)).filter(Boolean).join("\n\n");
}

function generateProvenance(
  state: ResearchState,
  sourceData: { paper_id: string; title: string }[],
  findingsData: Record<string, unknown>[],
  gaps: Gap[],
  contradictions: Contradiction[]
): string {
  const settings = getSettings();
  const now = new Date();
  const durationS = state.startedAt ? Math.floor((now.getTime() - state.startedAt.getTime()) / 1000) : 0;

  const verifiedCount = findingsData.filter((f) => f.is_verified).length;
  const fatalGaps = gaps.filter((g) => g.severity === "critical");
  const majorGaps = gaps.filter((g) => g.severity === "major");

  const dimLabels = state.researchDimensions.map((d) => d.label || d.strategy);

  const lines = [
    `# Provenance: ${state.query}`,
    "",
    "## Run metadata",
    `- **Date:** ${now.toISOString()}`,
    `- **Duration:** ${durationS}s`,
    `- **Model:** ${settings.model}`,
    `- **Rounds completed:** ${state.round} / ${state.maxRounds}${state.converged ? " *(converged early)*" : ""}`,
    `- **Research dimensions:** ${dimLabels.join(", ") || "default"}`,
    "",
    "## Source pipeline",
    `- **Sources consulted (pre-dedup):** ${state.papersConsulted || sourceData.length}`,
    `- **Sources accepted (post-filter):** ${sourceData.length}`,
    "",
    "## Findings",
    `- **Total findings:** ${findingsData.length}`,
    `- **Verified (confidence ≥ ${settings.confidenceThreshold}):** ${verifiedCount}`,
    `- **Unverified:** ${findingsData.length - verifiedCount}`,
    "",
    "## Reviewer integrity check",
    `- **FATAL gaps:** ${fatalGaps.length}`,
    `- **MAJOR gaps:** ${majorGaps.length}`,
    `- **Confirmed contradictions:** ${contradictions.length}`,
  ];

  if (contradictions.length > 0) {
    lines.push("", "### Contradictions");
    for (const c of contradictions) {
      const sev = c.similarity < 0.05 ? "FATAL" : c.similarity < 0.1 ? "MAJOR" : "MINOR";
      lines.push(`- [${sev}] ${c.description || "No description"}`);
    }
  }

  if (fatalGaps.length > 0 || majorGaps.length > 0) {
    lines.push("", "### Gap details");
    for (const g of [...fatalGaps, ...majorGaps]) {
      lines.push(`- [${g.severity}] ${g.description}`);
    }
  }

  const citationGraph = state.citationGraph;
  if (citationGraph && Object.keys(citationGraph).length > 0) {
    const inCorpusCitedBy: Record<string, number> = {};
    for (const citedList of Object.values(citationGraph)) {
      for (const citedId of citedList) {
        inCorpusCitedBy[citedId] = (inCorpusCitedBy[citedId] || 0) + 1;
      }
    }

    if (Object.keys(inCorpusCitedBy).length > 0) {
      const idToTitle: Record<string, string> = {};
      for (const s of sourceData) {
        idToTitle[s.paper_id] = s.title;
      }

      const topCited = Object.entries(inCorpusCitedBy)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 5);

      lines.push("", "## Top cross-cited papers (within corpus)");
      for (const [pid, count] of topCited) {
        const title = idToTitle[pid] || pid;
        lines.push(`- [${count} in-corpus citation${count > 1 ? "s" : ""}] **${title}**`);
      }
    }
  }

  lines.push("", "## Sub-questions");
  for (const sq of state.subQuestions) {
    lines.push(`- ${sq}`);
  }

  return lines.join("\n");
}

export async function writerNode(state: ResearchState): Promise<Partial<ResearchState>> {
  const settings = getSettings();
  const { query, findings, papers, startedAt } = state;
  const gaps = (state as unknown as { gaps?: Gap[] }).gaps || [];
  const contradictions = state.contradictions;

  const callback = state.progressCallback;
  if (callback) {
    callback("writer", { status: "writing" });
  }

  const [sourceData, sourceIndex] = buildSourceIndex(papers);

  const findingsData = findings.map((f) => ({
    finding_id: f.finding_id,
    claim: f.claim,
    confidence: f.confidence,
    confidence_label: confidenceLabel(f),
    is_verified: f.is_verified,
    source_paper_id: f.source_paper_id,
    citation_number: sourceIndex[f.source_paper_id] || "?",
  }));

  const [systemPrompt, userPrompt] = getWriterPrompt(query, findingsData, sourceData);

  let content: string;
  try {
    const client = getClient();
    content = await client.completeText(userPrompt, systemPrompt, { maxTokens: 4000 });
  } catch {
    content = generateFallbackReport(query, findings, papers);
  }

  const bibtex = generateBibtex(papers);

  let contradictionsMd = "";
  if (contradictions.length > 0) {
    const lines = ["## Contradictions Detected\n", "> The following claim pairs were identified as potentially contradicting each other.\n"];
    for (const c of contradictions) {
      const sev = c.similarity < 0.05 ? "FATAL" : c.similarity < 0.1 ? "MAJOR" : "MINOR";
      lines.push(`- **${sev}** — *"${c.claim_a}"* vs *"${c.claim_b}"*${c.description ? `\n  ${c.description}` : ""}`);
    }
    contradictionsMd = lines.join("\n");
  }

  const sections: Record<string, string> = { summary: content };
  if (contradictionsMd) {
    sections["contradictions"] = contradictionsMd;
  }

  const report = {
    report_id: `report-${Math.random().toString(36).slice(2, 10)}`,
    title: query,
    query,
    executive_summary: "",
    sections,
    sources: sourceData,
    source_index: sourceIndex,
    findings_summary: findingsData,
    bibliography: bibtex,
    created_at: new Date().toISOString(),
    status: "final" as const,
  };

  const outputDir = settings.outputDir;
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  fs.writeFileSync(path.join(outputDir, "report.md"), toMarkdown(report as any));
  fs.writeFileSync(path.join(outputDir, "report.json"), JSON.stringify(toJson(report as any), null, 2));
  fs.writeFileSync(path.join(outputDir, "report.bib"), bibtex);
  fs.writeFileSync(path.join(outputDir, "report.tex"), toLatex(report as any));

  const provenance = generateProvenance(state, sourceData, findingsData, gaps, contradictions);
  fs.writeFileSync(path.join(outputDir, "report.provenance.md"), provenance);

  return {
    report: report as any,
    status: "done" as const,
  };
}

export function createWriterNode() {
  return writerNode;
}